datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -1435,8 +1474,70 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
     }
 
     @Override
-    public void handleDataDeleted(String dataPath) throws Exception {
+    public void handleDataDeleted(String dataPath) {
       // do nothing
     }
   }
+
+  /**
+   * Listener for ZooKeeper state changes.
+   */
+  @VisibleForTesting
+  class ZkStateChangeListener implements IZkStateListener {
+    @Override
+    public void handleStateChanged(Watcher.Event.KeeperState state) {
+      LOG.info("ZkStateChangeListener:: handleStateChanged {}", state.toString());
+      switch (state) {
+        case Expired:
+          _timer.cancel();
+          onBecomeFollower();
+          exitOnSessionExpiryOrEstablishmentError();
+          return;
+        case Disconnected:
+          // if the session has expired it means that all the registration's ephemeral nodes are gone.
+          LOG.warn("Got {} event. Scheduling a system stop.", state.toString());
+          scheduleExpiryTimerAfterSessionTimeout();
+          return;
+        case SyncConnected:
+          LOG.info("ZkStateChangeListener::Connected. Canceling timer.");
+          _timer.cancel();
+          return;
+        default:
+          // Ignoring AuthFailed for now.
+      }
+    }
+
+    @Override
+    public void handleNewSession() {
+      LOG.info("ZkStateChangeListener:: A new session established.");
+    }
+
+    @Override
+    public void handleSessionEstablishmentError(final Throwable error) {
+      LOG.error("ZkStateChangeListener:: Failed to establish session.", error);
+      exitOnSessionExpiryOrEstablishmentError();
+    }
+  }
+
+  private void scheduleExpiryTimerAfterSessionTimeout() {
+    TimerTask timerTask = new TimerTask() {
+      public void run() {
+        exitOnSessionExpiryOrEstablishmentError();
+      }
+    };
+    _timer.schedule(timerTask, _sessionTimeoutMs);
+  }
+
+  // This is a temporary change to bring down the node on session expiry. We expect this to happen rarely.
+  // We are working on the transition of leader to follower on session expiry. On session expiry,
+  // the live instance (ephemeral nodes) gets deleted by the server and node automatically lose the election.
+  // By bringing down the node, the health monitoring services to restart the service is expected.
+  // This exit call will invoke the shutdownhook and will bring down the system. This is currently
+  // supported using a knob and adding shutdownhook will ensure the clean shutdown.
```

Comment:
Again, a bit confusing here. Not sure we need to add details about nurse bringing the node back up.  Also, not sure you really need to discuss the knob (which I assume is the config?) here. 
nit: shutdownhook -> shutdown hook

Maybe rephrase it to: Calling exit() will invoke the shutdown hook and bring down the system via a clean shutdown.

If you do want to mention the config, you can add a sentence: This behavior to bring down the system on session expiry is controlled via a config option.

If you do want to discuss the nurse job, you can add: A health monitoring service is expected to be in place to restart  the Brooklin service.

Reply:
ignoring, since not doing this now.